rtloader/common/builtins/datadog_agent.c

```diff
@@ -653,3 +662,49 @@ static PyObject *set_external_tags(PyObject *self, PyObject *args)
     Py_RETURN_NONE;
 
 }
+
+/*! \fn PyObject *obfuscate_sql(PyObject *self, PyObject *args)
+    \brief This function implements the `datadog_agent.obfuscate_sql` method, obfuscating
+    the provided sql string.
+    \param self A PyObject* pointer to the `datadog_agent` module.
+    \param args A PyObject* pointer to a tuple containing the key to retrieve.
+    \return A PyObject* pointer to the value.
+
+    This function is callable as the `datadog_agent.obfuscate_sql` Python method and
+    uses the `cb_obfuscate_sql()` callback to retrieve the value from the agent
+    with CGO. If the callback has not been set `None` will be returned.
+*/
+static PyObject *obfuscate_sql(PyObject *self, PyObject *args)
+{
+    // callback must be set
+    if (cb_obfuscate_sql == NULL) {
+        Py_RETURN_NONE;
+    }
+
+    PyGILState_STATE gstate = PyGILState_Ensure();
+
+    char *rawQuery;
+    if (!PyArg_ParseTuple(args, "s", &rawQuery)) {
+        PyGILState_Release(gstate);
+        return NULL;
+    }
+
+    char *obfQuery = NULL;
+    char *error_message = NULL;
+    obfQuery = cb_obfuscate_sql(rawQuery, &error_message);
+
+    PyObject *retval = NULL;
+    if (obfQuery != NULL) {
+        retval = PyStringFromCString(obfQuery);
+        cgo_free(obfQuery);
+    }
+    if (error_message != NULL) {
+        PyErr_SetString(PyExc_RuntimeError, error_message);
+        cgo_free(error_message);
+        // there shouldn't be any return value if there was an error, but null it out just in case
+        retval = NULL;
```

Comment:
If `obfQuery` wasn't `NULL` we just leaked a PyObject created line 698.

We should first check for error and `cgo_free(obfQuery)` (a NULL pointer will be discarded by cgo_free.
Then if not error, create the retval.

Reply:
Good catch! Updated